strawberryfields/api/result.py

```diff
@@ -0,0 +1,115 @@
+
+    * ``results.state``: The quantum state object contains details and methods
+      for manipulation of the final circuit state. Not available for remote
+      backends. See :doc:`/introduction/states` for more information regarding available
+      state methods.
+
+    * ``results.samples``: Measurement samples from any measurements performed.
+
+    **Example:**
+
+    The following examples run an existing Strawberry Fields
+    quantum :class:`~.Program` on the Gaussian engine to get
+    a results object.
+
+    Using this results object, the measurement samples
+    can be returned, as well as quantum state information.
+
+    >>> eng = sf.Engine("gaussian")
+    >>> results = eng.run(prog)
+    >>> print(results)
+    Result: 3 subsystems
+        state: <GaussianState: num_modes=3, pure=True, hbar=2>
+        samples: [0, 0, 0]
+    >>> results.samples
+    [0, 0, 0]
+    >>> results.state.is_pure()
+    True
+
+    .. note::
+
+        Only local simulators will return a state object. Remote
+        simulators and hardware backends will return
+        measurement samples  (:attr:`Result.samples`),
+        but the return value of ``Result.state`` will be ``None``.
+    """
+
+    def __init__(self, samples, is_stateful=True):
+        self._state = None
+        self._is_stateful = is_stateful
+
+        # ``samples`` arrives as a list of arrays, need to convert here to a multidimensional array
+        if len(np.shape(samples)) > 1:
+            samples = np.stack(samples, 1)
+        self._samples = samples
+
+    @property
+    def samples(self):
+        """Measurement samples.
+
+        Returned measurement samples will have shape ``(modes,)``. If multiple
+        shots are requested during execution, the returned measurement samples
+        will instead have shape ``(shots, modes)``.
+
+        Returns:
+            array[array[float, int]]: measurement samples returned from
+            program execution
+        """
+        return self._samples
+
+    @property
+    def state(self):
+        """The quantum state object.
+
+        The quantum state object contains details and methods
+        for manipulation of the final circuit state.
+
+        See :doc:`/introduction/states` for more information regarding available
+        state methods.
+
+        .. note::
+
+            Only local simulators will return a state object. Remote
+            simulators and hardware backends will return
+            measurement samples (:attr:`Result.samples`),
+            but the return value of ``Result.state`` will be ``None``.
+
+        Returns:
+            BaseState: quantum state returned from program execution
+        """
+        if not self._is_stateful:
+            raise AttributeError("The state is undefined for a stateless computation.")
+        return self._state
+
+    def __str__(self):
+        """String representation."""
+        return "Result: {} subsystems, state: {}\n samples: {}".format(
```

Comment:
Hmm, perhaps only smaller states could be outputted by this and there could be a separate auxiliary method that outputs every state.

That is, this could have a check along the lines of
```python
if len(self.state)>10 or len(self.samples)>1000:
# output lengths and advice users to use the Result.verbose_str option
# else proceed with outputting everything
```
Open to suggestions if this could work and if so, what the limit for these should be.

Reply:
Might be possible that the string-printing methods of numpy already take the size into account and truncate (haven't tested myself)

Reply:
Sorry, could've sworn I replied to this but maybe the GitHub outage ate it.

I tested this out, and it does seem that the samples get automatically truncated:

```python
In [4]: result = Result(np.random.randint((1e6, 8)))

In [6]: str(result)
Out[6]: 'Result: 1 subsystems, state: None\n samples: [[148373      4]]'
```

I should also note that this was a holdover from existing code that I simply moved to a different module, so I assume this was probably also discussed when it was first implemented?

Reply:
Oh, my bad, got confused with the syntax there :stuck_out_tongue: Thanks for checking that!

Reply:
@josh146 It seems like it's the `samples = np.stack(samples, 1)` that is taking time when printing the results (I'm using the master branch). This should be gone in this PR though, since it's an unnecessary line and apparently quite slow (which isn't too surprising since it not only transposes the array but also re-stacks all the inputs; I'm not sure why that was done that way though).

Reply:
Nice, good to know :+1: